elastic/beats
# filebeat: ID for filestream is required

## What does this PR do?
It makes mandatory to configure an ID to all filestream inputs. If an ID is added to a existing, working, configuration, no data is duplicated.

### How does it work
When a filestream input does not have an ID, we set the default `.global` ID at runtime and this ID is used to identify which files in the registry belong to each input.

Now that we're making the IDs mandatory, we need a way to migrate those entries in the registry to the newly set ID. Ideally this will happen only once when the users update to this new version of Filebeat.

They way we do it is by looking the registry for files that have their path matching the input but have the input ID as `.global`, those entries are then copied with the new key (which uses the input ID) so Filebeat can continue its operation without duplicating data.

Most of the logic is the same of the already existing method `UpdateIdentifiers` from the `sourceStore` but for this PR it was also needed to update the in-memory store right away. Also borrowing from `UpdateIdentifiers` usage, this is done during the file prospector initialisation.

## Why is it important?

## Checklist


~~## Author's Checklist~~
## How to test this PR locally
1. Create a simple config that reads a static file using `filestream`, do not set an ID for this input
2. Run Filebeat (a non patched version) and wait all events to be ingested
3. Stop Filebeat
4. Add an ID to your filestream input
5. Start a patched version of Filebeat
6. No new events should be ingested
7. The registry should have its entries updated with the new input ID 

## Related issues

- Closes #30061

~~## Use cases~~
~~## Screenshots~~
~~## Logs~~
